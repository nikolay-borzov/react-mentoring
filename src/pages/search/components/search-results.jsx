import React from 'react'
import PropTypes from 'prop-types'

import { GetLoadable } from '../../../components/helpers'

import { Radio, SearchResultsPanel } from '../../../components'

import { sortBy } from '../../../enums'

/* istanbul ignore next */
const FilmsGrid = GetLoadable(() =>
  import('../../../components/films-grid/films-grid')
)

export class SearchResults extends React.PureComponent {
  static propTypes = {
    foundCount: PropTypes.number.isRequired,
    displayCount: PropTypes.number.isRequired,
    films: PropTypes.arrayOf(PropTypes.object).isRequired,
    sortBy: PropTypes.oneOf(Object.values(sortBy)).isRequired,
    onSortByChange: PropTypes.func.isRequired
  }

  sortOptions = [
    {
      name: 'release date',
      value: sortBy.releaseDate
    },
    {
      name: 'rating',
      value: sortBy.rating
    }
  ]

  render() {
    const {
      sortBy,
      films,
      foundCount,
      displayCount,
      onSortByChange
    } = this.props

    const filmsFound = (
      <div>
        {foundCount}&nbsp;
        {foundCount > 1 ? 'movies' : 'movie'}
        &nbsp;found{' '}
        {foundCount > displayCount ? `(${displayCount} shown)` : null}
      </div>
    )

    return (
      <React.Fragment>
        <SearchResultsPanel>
          {filmsFound}
          <Radio
            name="sortBy"
            label="Sort by"
            value={sortBy}
            options={this.sortOptions}
            onChange={onSortByChange}
          />
        </SearchResultsPanel>
        <FilmsGrid films={films} />
      </React.Fragment>
    )
  }
}

export default SearchResults
