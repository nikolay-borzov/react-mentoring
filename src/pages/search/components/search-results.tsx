import React from 'react'

import { GetLoadable } from '../../../components/helpers'
import {
  SearchResultsPanel,
  RadioControlled,
  FilmsGridProps
} from '../../../components'
import { sortBy } from '../../../enums'
import { Film } from '../../../entities/film'
import { SortBy } from '../../../services/film-service'

/* istanbul ignore next */
const FilmsGrid = GetLoadable<FilmsGridProps>(() =>
  import('../../../components/films-grid/films-grid')
)

export interface SearchResultsProps {
  foundCount: number
  displayCount: number
  films: Film[]
  sortBy?: SortBy
  onSortByChange(value: string): void
}

export class SearchResults extends React.PureComponent<SearchResultsProps> {
  sortOptions = [
    {
      name: 'release date',
      value: sortBy.releaseDate
    },
    {
      name: 'rating',
      value: sortBy.rating
    }
  ]

  render() {
    const {
      sortBy,
      films,
      foundCount,
      displayCount,
      onSortByChange
    } = this.props

    const filmsFound = (
      <div>
        {foundCount}&nbsp;
        {foundCount > 1 ? 'movies' : 'movie'}
        &nbsp;found{' '}
        {foundCount > displayCount ? `(${displayCount} shown)` : null}
      </div>
    )

    return (
      <>
        <SearchResultsPanel>
          {filmsFound}
          <RadioControlled
            name="sortBy"
            label="Sort by"
            value={sortBy}
            options={this.sortOptions}
            onChange={onSortByChange}
          />
        </SearchResultsPanel>
        <FilmsGrid films={films} />
      </>
    )
  }
}

export default SearchResults
