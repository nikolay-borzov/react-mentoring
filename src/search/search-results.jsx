import React from 'react'
import PropTypes from 'prop-types'

import { SearchResultsPanel } from './search-results-panel'
import { Radio, FilmGrid } from '../core/components'

import { sortBy } from '../core/enums'

export class SearchResults extends React.Component {
  static propTypes = {
    foundCount: PropTypes.number.isRequired,
    films: PropTypes.arrayOf(PropTypes.object).isRequired,
    sortBy: PropTypes.oneOf(Object.values(sortBy)),
    onSortByChange: PropTypes.func.isRequired
  }

  constructor(props) {
    super(props)

    this.sortOptions = [
      {
        name: 'release date',
        value: sortBy.releaseDate
      },
      {
        name: 'rating',
        value: sortBy.rating
      }
    ]
  }

  render() {
    const filmsFound = (
      <div>
        {this.props.foundCount}&nbsp;
        {this.props.foundCount > 1 ? 'movies' : 'movie'}
        &nbsp;found
      </div>
    )

    return (
      <React.Fragment>
        <SearchResultsPanel>
          {filmsFound}
          <div className="flex">
            <Radio
              name="sortBy"
              label="Sort by"
              value={this.props.sortBy}
              options={this.sortOptions}
              onChange={this.props.onSortByChange}
            />
          </div>
        </SearchResultsPanel>
        <FilmGrid {...this.props} />
      </React.Fragment>
    )
  }
}
