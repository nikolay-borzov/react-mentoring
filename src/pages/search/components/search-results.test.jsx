import React from 'react'

import { SearchResults } from './search-results'

describe('SearchResults component', () => {
  let props = {
    onSortByChange: jest.fn()
  }

  beforeEach(() => {
    props.foundCount = 3000
    props.displayCount = 15
    props.films = []
    props.sortBy = 'release_date'

    props.onSortByChange.mockClear()
  })

  it('renders correctly', () => {
    const wrapper = shallow(<SearchResults {...props} />)

    expect(wrapper).toMatchSnapshot()
  })

  it('renders correctly when only one movie is found', () => {
    props.foundCount = 1
    props.displayCount = 1

    const wrapper = shallow(<SearchResults {...props} />)

    expect(wrapper).toMatchSnapshot()
  })

  it('renders correctly when found count is bigger than limit value', () => {
    props.foundCount = 10
    props.displayCount = 11

    const wrapper = shallow(<SearchResults {...props} />)

    expect(wrapper).toMatchSnapshot()
  })
})
