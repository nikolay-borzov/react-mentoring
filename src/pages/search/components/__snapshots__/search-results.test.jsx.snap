// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResults component renders correctly 1`] = `
<React.Fragment>
  <SearchResultsPanel>
    <div>
      3000
       
      movies
       found (
      15
       shown)
    </div>
    <Radio
      label="Sort by"
      name="sortBy"
      onChange={[MockFunction]}
      options={
        Array [
          Object {
            "name": "release date",
            "value": "release_date",
          },
          Object {
            "name": "rating",
            "value": "vote_average",
          },
        ]
      }
      style="plain"
      value="release_date"
    />
  </SearchResultsPanel>
  <FilmsGrid
    films={Array []}
  />
</React.Fragment>
`;

exports[`SearchResults component renders correctly when only one movie is found 1`] = `
<React.Fragment>
  <SearchResultsPanel>
    <div>
      1
       
      movie
       found (
      1
       shown)
    </div>
    <Radio
      label="Sort by"
      name="sortBy"
      onChange={[MockFunction]}
      options={
        Array [
          Object {
            "name": "release date",
            "value": "release_date",
          },
          Object {
            "name": "rating",
            "value": "vote_average",
          },
        ]
      }
      style="plain"
      value="release_date"
    />
  </SearchResultsPanel>
  <FilmsGrid
    films={Array []}
  />
</React.Fragment>
`;
